components/cloud/aws/src/kms.rs

```diff
@@ -0,0 +1,262 @@
+        config: Config,
+        dispatcher: Dispatcher,
+        credentials_provider: Creds,
+    ) -> Result<AwsKms>
+    where
+        Creds: ProvideAwsCredentials + Send + Sync + 'static,
+        Dispatcher: DispatchSignedRequest + Send + Sync + 'static,
+    {
+        let region = util::get_region(
+            config.location.region.as_ref(),
+            config.location.endpoint.as_ref(),
+        )?;
+        let client = KmsClient::new_with(dispatcher, credentials_provider, region);
+        Ok(AwsKms {
+            client,
+            current_key_id: config.key_id,
+            region: config.location.region,
+            endpoint: config.location.endpoint,
+        })
+    }
+
+    fn new_with_dispatcher<D>(config: Config, dispatcher: D) -> Result<AwsKms>
+    where
+        D: DispatchSignedRequest + Send + Sync + 'static,
+    {
+        let credentials_provider = util::CredentialsProvider::new()?;
+        Self::new_creds_dispatcher(config, dispatcher, credentials_provider)
+    }
+
+    pub fn new(config: Config) -> Result<AwsKms> {
+        let dispatcher = util::new_http_client()?;
+        Self::new_with_dispatcher(config, dispatcher)
+    }
+}
+
+#[async_trait]
+impl KmsProvider for AwsKms {
+    fn name(&self) -> &[u8] {
+        AWS_KMS_VENDOR_NAME
+    }
+
+    // On decrypt failure, the rule is to return WrongMasterKey error in case it is possible that
+    // a wrong master key has been used, or other error otherwise.
+    async fn decrypt_data_key(&self, data_key: &EncryptedKey) -> Result<Vec<u8>> {
+        let decrypt_request = DecryptRequest {
+            ciphertext_blob: bytes::Bytes::copy_from_slice(&*data_key),
+            // Use default algorithm SYMMETRIC_DEFAULT.
+            encryption_algorithm: None,
+            // Use key_id encoded in ciphertext.
+            key_id: Some(self.current_key_id.deref().clone()),
+            // Encryption context and grant tokens are not used.
+            encryption_context: None,
+            grant_tokens: None,
+        };
+        self.client
+            .decrypt(decrypt_request.clone())
+            .await
+            .map_err(classify_decrypt_error)
+            .map(|response| response.plaintext.unwrap().as_ref().to_vec())
+    }
+
+    async fn generate_data_key(&self) -> Result<DataKeyPair> {
+        let generate_request = GenerateDataKeyRequest {
+            encryption_context: None,
+            grant_tokens: None,
+            key_id: self.current_key_id.deref().clone(),
+            key_spec: Some(AWS_KMS_DATA_KEY_SPEC.to_owned()),
+            number_of_bytes: None,
+        };
+        self.client
+            .generate_data_key(generate_request)
+            .await
+            .map_err(classify_generate_data_key_error)
+            .and_then(|response| {
+                let ciphertext_key = response.ciphertext_blob.unwrap().as_ref().to_vec();
+                let plaintext_key = response.plaintext.unwrap().as_ref().to_vec();
```

Comment:
avoid unwrap and handle the error.

Reply:
This code was just moved from encryption/master_key/kms.rs and is unchanged.

Reply:
and we can change them by the way.

Reply:
I know, I have changed so much code already. I need to stop refactoring this code now and instead commit it.